Add tabbed mode spec

mgiuca opened the pull request:
Based on an earlier draft by @loubrett (Louise Brett).


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/mgiuca/manifest-incubations/pull/95.html" title="Last updated on Apr 26, 2024, 5:52 AM UTC (57014a6)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/manifest-incubations/95/db88c85...mgiuca:57014a6.html" title="Last updated on Apr 26, 2024, 5:52 AM UTC (57014a6)">Diff</a>

mgiuca committed:
Added tabbed to candidate display mode list.

mgiuca committed:
Update note.

mgiuca committed:
Added tabbed display mode and tab_strip member.

mgiuca committed:
Added full details on the tab_strip member.

Copy-pasted from Louise Brett's source doc. Need to clean up, linkify and edit.

mgiuca committed:
Linkify concepts.

mgiuca committed:
Format processing steps.

mgiuca committed:
Fix up links and add notes to be fixed up.

mgiuca committed:
Fix construction of URL pattern. Adds a TODO with more questions.

mgiuca committed:
Use manifest URL instead of scope URL as the base for processing scope patterns.

mgiuca committed:
Linkify list.

mgiuca committed:
Add link to manifest PR to linkify manifest and start URL.

mgiuca committed:
Clean up warnings.

mgiuca commented:
@loubrett Ready for you to look now.

mgiuca committed:
Added a TODO about url.

loubrett commented on index.html:
```diff
@@ -84,76 +84,87 @@ <h2>
+            <dfn data-export="" data-dfn-for="display mode">tabbed</dfn>
+          </dt>
+          <dd>
+            The web application can have multiple application contexts combined
```
Should 'application context' here be a link? Same for the other one in this paragragh and more at line 251 and 252.

mgiuca replied:
Done. I won't repeatedly link multiple times per paragraph.

loubrett commented on index.html:
```diff
@@ -186,31 +197,272 @@ <h4>
+        </aside>
+        <p>
+          If the [=tab_strip/home_tab=] member is not present, then the
+          application is not intended to have a home tab context, otherwise it
```
Should 'home tab context' here be a link? Same for line 301

mgiuca replied:
Done.

loubrett commented on index.html:
```diff
@@ -186,31 +197,272 @@ <h4>
+        <p>
+          The <dfn>home tab scope</dfn> is the set of URLs to which the home
+          tab context can be navigated to. All navigations within the
+          application to a URL within the home tab scope will be done in the
```
Should 'home tab scope' here be a link?

mgiuca replied:
Done.

loubrett commented on index.html:
```diff
@@ -186,31 +197,272 @@ <h4>
+        </ul>
+        <p>
+          If [=new_tab_button/url=] is within the [=home tab scope=], then the
+          new tab button may be hidden.
```
Should 'may' be capitalized?

mgiuca replied:
Good call-out. I think it's actually good to strengthen this to a MUST.

And further, I think this rule should be moved to the parser: instead of simply saying "don't show the button in the UI", the parser should just ignore it if it's within scope. Would that change the behaviour at all? (I don't think so, although internally it would mean we are discarding the illegitimate new_tab URL rather than storing it in the app database, which is slightly risky if we want to change this rule later.)

loubrett replied:
Yeh we could change the parser to ignore it.
Currently later in the code, if the new tab URL is unset, we use the start_url (which will then always be in the home tab scope so it gets hidden). But we could leave it unset, and make sure the parser sets it to start_url when relevant.
I don't know how likely it is to want to change this rule, seems like very strange behaviour if we allowed it.

loubrett reviewed (commented):
Just some minor comments about linking stuff.

I haven't thought through the URLPattern stuff any more yet so I'll have a think about that next week.

mgiuca committed:
Respond to review.

mgiuca committed:
Moved processing sections into subsections.

Now there is just a single processing section at the h2 level.

mgiuca committed:
Major rewrite of tab_strip section.

Too numerous to list individual changes. A lot of moving around.

Many additional details added, including:
- List of invariants that must be followed (e.g. home tab never
  navigates out of home scope, regular tab never navigates into home
  scope).
- Specifying that home tabs open at the start_url.
- Properly defining within and outside of home scope terms.
- Properly defining what happens when you click the new tab button.

mgiuca commented:
I just pushed a major rework. Still incomplete so don't re-review yet.

mgiuca committed:
Better define home tab context and home tab scope.

mgiuca committed:
scope_patterns: Scope pattern can now be either a full dict or a string.

mgiuca committed:
Define has a home tab, for more clarity.

mgiuca committed:
Minor refactor.

mgiuca committed:
Move the check of the new tab button being outside of home tab scope into the parser.

mgiuca committed:
Added a note about History API.

mgiuca committed:
Use 'has a home tab'.

mgiuca committed:
Added usage example.

mgiuca committed:
Fix link to processed manifest.

mgiuca committed:
Use build a URL pattern instead of hand-creating a dictionary.

mgiuca committed:
Invoke process the tab_strip member.

mgiuca committed:
Properly define new_tab_button/url.

mgiuca committed:
Remove issue about missing exports - now resolved.

mgiuca committed:
Added link to at-risk issue 97 (parsing scope patterns as a string).

mgiuca committed:
Refactor definition of home tab scope.

Define a boolean condition first, rather than defining a set. Makes it more readable.

Also exclude fragments (but not query).

mgiuca committed:
exclude URLs outside of manifest scope from being in home scope.

mgiuca committed:
Remove the official definition of home tab scope, as it is only used non-normatively, and just talk about it as an idea, not a formal term.

Add a note about ignoring the fragment but not the query.

mgiuca committed:
Update usage example.

Do not use the at-risk string syntax.

Add more detail about the default matching the query, or not.

mgiuca committed:
Rewrote invariants concerning home tab scope.

Now, the three invariants are non-normative discussion of what is intended, rather than a hard rule. The normative logic elsewhere should speak for itself.

This way, we do not need to formally find a concept of 'initial URL' and can just informally talk about documents that rewrite their URL.

mgiuca commented:
Hi @loubrett . This is finally ready for review. PTAL; note that you can preview it [here](https://pr-preview.s3.amazonaws.com/mgiuca/manifest-incubations/pull/95.html).

mgiuca commented:
@dmurph adding as editor. PTAL, thanks, and let me know if you want to discuss!

loubrett commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+        </p>
+        <p>
+          How the [=home tab context=] is presented is at the discretion of the
+          user agent, but it SHOULD have a different appearance to normal tabs.
```
Should this say `normal application contexts` instead of `normal tabs`? You don't really use the word tab anywhere else.

mgiuca replied:
Done.

loubrett commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+        <p>
+          An application <dfn>has a home tab</dfn> if the [=Document/processed
+          manifest=] includes a non-null [=home_tab=] member of the
+          [=tab_strip=] member.
```
Should the `has a home tab` definition also  say the display mode needs to be tabbed? Or is it implied that none of these fields are relevant in that case?

mgiuca replied:
Hmm, yes, I think you're right. It's a little tricky since technically applications don't have display modes, application contexts do. So it's a little wrong to say "if the application's display mode is X" but I think we just need to go with that since that clearly shows the intention.

We do need something like this because later we say "whenever a new application window is created ... the user agent MUST create a new home tab context" which clearly can't be true for non-tabbed apps.

loubrett commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+            tab context=], and all navigations to a URL [=outside of home tab
+            scope=] will be performed in a different [=application context=]
+            (i.e., another tab), creating one if necessary. The application's
+            [=start URL=] is always part of the home tab scope.
```
Nit (and probably doesn't matter), but here you say the start URL is always in the home tab scope but then below you give a case where it isn't (because the scope is the empty set).

mgiuca replied:
"and probably doesn't matter" --> it's a web standard, everything matters! :joy:

dmurph commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+        </section>
+        <section>
+          <h3>
+            Navigations concerning the home tab scope
```
As per my exploration for link capturing, this has potential to break sites. I recommend linking the 'navigating' keyword here into somewhere in the html spec. Likely in 

Specifically, I recommend probably saying something along the lines of "When [creating a new top-level browsing context and document](https://html.spec.whatwg.org/multipage/document-sequences.html#creating-a-new-top-level-browsing-context),  or [creating a new auxilary browsing context and document](https://html.spec.whatwg.org/multipage/document-sequences.html#creating-a-new-auxiliary-browsing-context), choose the 'home' tab's browsing context when [creating a new browsing context](https://html.spec.whatwg.org/multipage/document-sequences.html#creating-a-new-browsing-context)." (that's actually probably not great to add the 'conditional execution of algo C if you are in algo A or B - perhaps we can just modify C w/o the condition? But this is probably enough to explain my request).

Or -perhaps you'll want to say something like "create a new browsing context that replaces the home tab browsing context".

Example of where this can cause error - if you have named frames, and you open a non-home-tab url (foo.com/away, named frame "B"), and then a user clicks a link that navigates the "B" frame to foo.com, which is in the home tab context.

Or - if the user clicks on a navigate-self link in a tab that is not the home tab, but navigates into the home tab context, what do you want?

It would be really helpful to reference the spec here around navigations, and this helps prevent things like breakages with window.opener and window.frames - the way this is implemented it has to be done where the browsing context gets chosen or created for the navigation in order to not break that.

mgiuca replied:
I'll need to block out some more time to think about this... starting off, adding a link to "navigate".

I think it's OK to just hand-wavily say "if navigating from...", after all that's roughly what the definition of NavigateEvent does ([link](https://html.spec.whatwg.org/multipage/nav-history-apis.html#the-navigate-event)).

dmurph replied:
If you can, I think it would be best to hook this into the navigation algorithm. I think that should be possible - I think the correct behavior here is to say that when a navigation occurs within an application context that is 'tabbed', if that navigation [creates a new browsing context](https://html.spec.whatwg.org/multipage/document-sequences.html#creating-a-new-browsing-context), then it should use the 'home tab' browsing context.

What do you think?

dmurph commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+        </section>
+        <section class="informative">
+          <h3>
+            Home tab invariants
```
These invariants are likely 'cleaned up' or explained better by my above suggestion about exactly where the 'home tab context' is chosen to be used for a navigation. Then this is more of an informative note / explanation of the behavior, and the edge cases are:
- window.opener (actually this might not be a huge deal - you can just choose the home browsing context for the navigation & allow that to be part of the browsing group the link came from)
- frame name is specified, which already exists
- self navigation (kinda the same as frame name)
- the history modification API as you mentioned
- others?

mgiuca replied:
This section is intended to be informative. I intend for this to be (non-normatively) the overriding philosophy of how tabs work, while the normative text makes sure that it happens.

I'll have to think about window.opener and frames which I hadn't previously considered. However I think frames don't matter because this only applies to top-level browsing contexts.

mgiuca replied:
We had an offline discussion about this. The summary is:

1. The paragraph that explains to developers of single-page applications that they should do real navigations in and out of home scope, is to be reworded so it's clear that it's giving developers advice, rather than highlighting an issue.
2. The "navigations concerning the home tab scope" should be reworked (and we workshopped a sketch of how the algorithm should look, which is as follows):

> When [choosing a navigable for a navigation ](https://html.spec.whatwg.org/multipage/document-sequences.html#the-rules-for-choosing-a-navigable) from the [home tab context](https://pr-preview.s3.amazonaws.com/mgiuca/manifest-incubations/pull/95.html#dfn-home-tab-context) to a [URL](https://url.spec.whatwg.org/#concept-url) url that is [outside of home tab scope](https://pr-preview.s3.amazonaws.com/mgiuca/manifest-incubations/pull/95.html#dfn-outside-of-home-tab-scope), the following steps are run:
>
> 1. Behave as `target=_blank`, with _noopener_ and _noreferrer_ specified, creating a new top-level traversable _t_.
t is created in the current window and has the manifest applied.
> 2. Focus _t_.
>
> When [choosing a navigable for a navigation](https://html.spec.whatwg.org/multipage/document-sequences.html#the-rules-for-choosing-a-navigable), if the [URL](https://url.spec.whatwg.org/#concept-url) url is [within home tab scope](https://pr-preview.s3.amazonaws.com/mgiuca/manifest-incubations/pull/95.html#dfn-within-home-tab-scope), the following steps are run:
>
> 1. Let _t_ be the home tab context’s traversable associated with the current window.
> 2. Override the [algorithm](https://html.spec.whatwg.org/multipage/document-sequences.html#the-rules-for-choosing-a-navigable) to return _t_.
> 3. Focus _t_.

mgiuca replied:
@dmurph 

A couple of issues arose trying to implement this:

1. It turns out that "[choosing a navigable](https://html.spec.whatwg.org/multipage/document-sequences.html#the-rules-for-choosing-a-navigable)" is not exported, so for whatever reason, we are not supposed to reference it from outside of HTML. I'm not sure if that means we just shouldn't, or should try and export it, but to avoid shaving that yak first, I think I will just hand-wave it instead of directly referencing this algorithm.
2. We don't actually want to always override the algorithm, only when target=_self. (So for instance, if you click a link with target=<name> then it should still open in the named target.) However, conversely, we want to capture navigations that target this tab, even if they didn't come from this tab (for example, if a child frame does a navigation with target=_parent, into a home URL, we want to capture that into the home tab). So actually I think hooking into the "choosing a navigation" is not the right place. We should let that algorithm run its course, and then _after_ a navigable has been chosen, if it is one of our tabs, override the navigable that actually gets navigated.

I think on account of both of these, it makes more sense to go back to hooking into the algorithm at the "navigate" step. I'll rework it and send it for review.

mgiuca replied:
I've rewritten it to capture the spirit of what we talked about this morning.

I am using "choose a navigable" to get the target=_blank logic, with a red box noting that I can't actually access it (#98).

But primarily, this is now about intercepting "navigation" and doing an "instead of navigating X, navigate Y" type of logic. I think that is more precisely what we want, because then we will intercept if _anything_ in the system tries to navigate one of our tabs.

dmurph replied:
LGTM

dmurph commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+          steps are run:
+        </p>
+        <ol>
+          <li>Create a new [=application context=] in the current window and
```
perhaps reference the HTML spec algo too somehow? https://html.spec.whatwg.org/multipage/document-sequences.html#creating-a-new-browsing-context

mgiuca replied:
Actually there is an existing definition of [create a new application context](https://www.w3.org/TR/appmanifest/#dfn-create-a-new-application-context) which I can leverage here.

dmurph commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+          </h3>
+          <p>
+            To <dfn>process the `home_tab` member</dfn>, given [=ordered map=]
+            |json:ordered map|, [=ordered map=] |manifest tab strip:ordered
```
perhaps name this `json_tab_strip`  to help readability?

mgiuca replied:
Done.

dmurph commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+          </h3>
+          <p>
+            To <dfn>process the `scope_patterns` member</dfn>, given [=ordered
+            map=] |json:ordered map|, [=ordered map=] |manifest home
```
same? where are we in the json of the manifest?

mgiuca replied:
Done (we're in home_tab/scope_patterns).

dmurph commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+          </h3>
+          <p>
+            To <dfn>process the `home_tab` member</dfn>, given [=ordered map=]
+            |json:ordered map|, [=ordered map=] |manifest tab strip:ordered
```
maybe rename this here and elsewhere `manifest_tab_strip_member`?

Or- not sure what is 'normal' for spec - just wanted to make it clearer this was a variable that is in the parsed manifest output. Fine to keep as-is if it matches convension.

mgiuca replied:
Unclear if this is following any other established pattern. (Most of the "complex" processing steps I can see do not take a manifest member, they just return one as a result. We could do that here, but this way is more flexible and follows the top-level better, so I will choose to keep it unless you object.

It is quite normal to have variable names (in spec) with spaces in them, not underscores.

dmurph commented on index.html:
```diff
@@ -218,6 +237,400 @@ <h2>
+        </aside>
+        <section>
+          <h3>
+            Every window has a home tab
```
There are some assumptions in this text that the browsing contxt is in `tabbed` display mode. The definition of the `application context` doesn't differentiate windowed vs in the browser. It might be good to clarify that the context for some of these assertions is that the window is in tabbed display mode.... you may need to define a mulit-browser-context window? Not sure if that is in spec somewhere.

mgiuca replied:
Louise just raised this: I have amended the definition of "has a home tab" to require that the display mode is tabbed, which I think addresses this.

> The definition of the application context doesn't differentiate windowed vs in the browser.

I'm pretty sure it does (though not very clearly) - the definition of "[application context](https://www.w3.org/TR/appmanifest/#dfn-application-context)" is "A top-level browsing context that has a manifest applied to it" which my understanding means an application window. Otherwise I think a lot of our assumptions throughout the Manifest spec are flawed.

dmurph reviewed (commented):
Some initial comments - also happy to go over in-person in a meeting

mgiuca committed:
Fix the logic of whether the app has a new tab button.

Now has an explicit concept of "has a new tab button" (analogous to "has
a home tab"). The parser algorithm no longer (accidentally) sets the new
tab URL to the start URL if it is within home tab scope.

Add a note clarifying why the new tab button is not allowed to be in
home tab scope.

Also note the start URL default, which is easy to miss as it is buried
in the parser algorithm.

mgiuca committed:
Respond to loubrett review.

mgiuca committed:
Some basic response to dmurph review.

mgiuca committed:
Rewrote confusing advice for developers.

mgiuca committed:
Rewrite navigations concerning the home tab scope.

Done in consultation with dmurph and loubrett, to make the navigation rules much clearer and hook into the HTML navigable machinery.

mgiuca committed:
Link to a GitHub issue instead of describing the issue inline.

mgiuca committed:
Following dmurph's advise to standardize the variable names in processing steps.

mgiuca commented:
@dmurph @loubrett Finished responding to both of your reviews. Thanks for the insightful comments & meeting this morning.

See my comments above - I was not able to do literally what we discussed but I think I have captured the behaviour you wanted. PTAL.

dmurph reviewed (approved):
LGTM, although nit: the display_mode link in the rendered text seems broken.

loubrett reviewed (approved):
LGTM

mgiuca commented:
> LGTM, although nit: the display_mode link in the rendered text seems broken.

Thanks Dan.

Yes, I mentioned this to you in a chat: we need to wait until w3c/csswg-drafts#7306 gets pushed out to the editor's draft, and then it will automatically resolve itself (because the definition of "display mode" was moved from CSS mediaqueries-5 to Manifest; Manifest updated right away, CSS did not). I don't know when that will be; it was last updated in May 2022, coming up to two years...
